### ISSUE TITLE
Feature Request (Writable Object)

### ISSUE THREAD

**Issue by 0xGreen** [CravateRouge/bloodyAD#52]
If i am not missing something then it is only possible to get the writable object for the current used account, For Example:

```bash
bloodyAD --host dc.test.com -d test.com -u user1 -p StrongPass get writable 

# distinguishedName: CN=S-1-5-11,CN=ForeignSecurityPrincipals ...
# permission: WRITE

# distinguishedName: CN=user1 ...
# permission: WRITE
```

Now, imagine a scenario, where user *user1* is member of a group *MgtUser*  and *MgtUser* has *ExtendedWrite* or *Force Change Password* etc rights on user *user2*.

There was no way to find writable objects for group *MgtUser*. [ User1 > MgtUser > ExtendedRight on User2]

So, if possible some like membership feature:

```bash
bloodyAD --host dc.test.com -d test.com -u user1 -p StrongPass get membership user1
bloodyAD --host dc.test.com -d test.com -u user1 -p StrongPass get membership user2
```

Maybe something similar (just an idea)

```bash
bloodyAD --host dc.test.com -d test.com -u user1 -p StrongPass get writable user1
bloodyAD --host dc.test.com -d test.com -u user1 -p StrongPass get writable MgtUser
```

Hope the explanation make sense.

Thanks, G.

**Comment by ozelis** [comment]
I have implemented this in my own fork in `https://github.com/ozelis/bloodyAD/commit/b8ec2f701ba21c39236a4a41a0e593289ce4c7ed`. 
I added a new `rights()` function in `cli_modules/get.py` that returns rights for a given target. It works something like this:
```bash
$ bloodyAD --host REDACTED -s -d 'REDACTED.HTB' -u 'redacted' -p 'redacted' get rights user1

distinguishedName: CN=redacted,CN=Users,DC=redacted,DC=htb
User-Force-Change-Password: CONTROL_ACCESS


$ bloodyAD --host REDACTED -s -d 'REDACTED.HTB' -u 'redacted' -p 'redacted' get rights user2

distinguishedName: DC=REDACTED,DC=htb
DS-Replication-Get-Changes-In-Filtered-Set: CONTROL_ACCESS
DS-Replication-Get-Changes: CONTROL_ACCESS
DS-Replication-Synchronize: CONTROL_ACCESS
DS-Replication-Get-Changes-All: CONTROL_ACCESS


$ bloodyAD --host REDACTED -s -d 'REDACTED.HTB' -u 'redacted' -p 'redacted' get rights Administrator

distinguishedName: DC=redacted,DC=htb
DS-Replication-Get-Changes-In-Filtered-Set: CONTROL_ACCESS
DS-Replication-Get-Changes: CONTROL_ACCESS
DS-Replication-Synchronize: CONTROL_ACCESS
DS-Replication-Manage-Topology: CONTROL_ACCESS
DS-Replication-Get-Changes-All: CONTROL_ACCESS
Read-Only-Replication-Secret-Synchronization: CONTROL_ACCESS

distinguishedName: CN=Builtin,DC=redacted,DC=htb
DS-Replication-Get-Changes-In-Filtered-Set: CONTROL_ACCESS
DS-Replication-Get-Changes: CONTROL_ACCESS
DS-Replication-Synchronize: CONTROL_ACCESS
DS-Replication-Manage-Topology: CONTROL_ACCESS
DS-Replication-Get-Changes-All: CONTROL_ACCESS
Read-Only-Replication-Secret-Synchronization: CONTROL_ACCESS

distinguishedName: CN=DC01,OU=Domain Controllers,DC=redacted,DC=htb
DS-Validated-Write-Computer: WRITE_VALIDATED
```

I can submit a PR if you're interested or just use that as you seem fit.

**Comment by CravateRouge** [comment]
Hey @ozelis , you're suggestion is very interesting! I think we could merge `get rights` and `get writable` into the same function. If user requested is same as the user used to connect to ldap we can use allowedChildClassesEffective and allowedAttributesEffective and if it's another one use your code. Also it would be interesting to keep the same nomenclature putting `permission: <permission_type>` or if `--detail`: `<attribute_name>: <permission_type>`.
And finally do it in a transitive (on all domain we can connect to) manner as you can see in `dnsDump`:
```
if transitive:
        trustmap = conn.ldap.getTrustMap()
        for trust in trustmap.values():
            if "conn" in trust:
```

**Comment by ozelis** [comment]
I think I  misunderstood the original Issue. They wanted for `get writable` to work for an arbitrary user and this is a little different as it shows  all *extended* rights some user has on objects.

Here `get writable` would return objects/attributes we can write to, but having some [extended right](https://learn.microsoft.com/en-us/windows/win32/adschema/extended-rights) does not (necessary) mean we can write to some attribute of that object. For example `DS-Replication-Synchronize`  means you can do "dcsync" but that is done through entirely different protocol ([MS-DRSR](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-drsr/f977faaa-673e-4f66-b9bf-48c640241d47)). 

The only case I can think of is that if we have `User-Force-Change-Password` right that means we can write to `unicodePwd` attribute, although this is not reported in `get writable` even with that right. Here it might make some sense to specifically check for  `User-Force-Change-Password` during `get writable` to report that `uncidePwd` is writable.

To solve the original issue we'd  have to parse `nTSecurityDescriptor` for each object and check if target user SID (or SID of any groups that user is a member of) has GenericWrite on it. And if user does not have GenericAll on the object it might still have WriteProperty on *some* attributes. 

I'll look into this some more. But I think it would be best to leave `get rights` for extended rights as is, and then change `get writable` to check for writable attributes from `nTSecurityDescriptor` if different user is specified as an optional argument. What do you think?

**Comment by ozelis** [comment]
I'm playing around with this now and it's not that simple to do the same thing as ldap does automatically with `sDRightsEffective`/`allowedAttributesEffective`/`allowedChildClassesEffective`. For my use case I was just checking for right guids that are in the `adschema.CONTROL_ACCESS_RIGHTS` as the only other way to get this information was to query individual objects with `get object --resolve-sd` and then try to eyeball the `nTSecurityDescriptor` dump to figure out what users/group have interesting rights.   

Trying to replicate the functionality of `get writable` I'm running into some issues: 

- Instead of granting rights to individual attributes, it sometimes does this for [property sets](https://learn.microsoft.com/en-us/windows/win32/adschema/property-sets) and I'm not sure if i should hard-code which attributes are in a property set (and they are different for different versions of windows to make it more confusing) or maybe there is a way to query ldap schema to figure this out.
- Generally ACE looks something like `(mask, rights_guid, trustee_sid)` but sometimes `trustee_sid` is one of the special RIDs like [`SECURITY_PRINCIPAL_SELF_RID (S-1-5-10)`](https://learn.microsoft.com/en-us/windows/win32/secauthz/well-known-sids) and I'm supposed to treat it as the SID of the object I'm querying, but then if the object is a group the meaning of that ACE is that members of that group have that right on their own object (?), not the group object itself and it reports incorrectly that these attributes are writable on the group object. 
- Then there's the whole deal with the inherited permissions and the way they resolved that I have trouble understanding. I'll have to read more about this or maybe look how utilities that do collection for BloodHound deal with this.

I'll try cook something up though and see how it goes.

**Comment by CravateRouge** [comment]
You're right it's not that straightforward but it's doable and you can see how it works as bloodyAD already does it when parsing the NTSecurityDescriptor with `get object --resolve-sd`. The code can be a little confusing sometimes because of the resolution lazyness but everything is there and you also have the [wiki ](https://github.com/CravateRouge/bloodyAD/wiki/Access-Control) to explain it more in details if you want to dig a little into it :)
